cmd/generate.go

```diff
@@ -83,26 +84,59 @@ func genProcessCluster(cmd *cobra.Command, clusterName string, p *ants.Pool) {
 
 	// determine list of components to process
 	var compList []string
+	var currentCompList []string
+
 	if components != "" {
 		// only process specified component if it's defined in the cluster
-		for c, _ := range clusterComponents {
-			for _, b := range strings.Split(components, ",") {
+		for _, b := range strings.Split(components, ",") { //flipped so it could append to currentCompList first
+			currentCompList = append(currentCompList, b)
+			for c, _ := range clusterComponents {
 				matched, _ := regexp.MatchString("^"+b+"$", c)
 				if matched {
 					compList = append(compList, c)
 				}
 			}
 		}
-		if len(compList) == 0 {
-			return
-		}
 	} else {
 		for c, _ := range clusterComponents {
 			compList = append(compList, c)
 		}
+		// list of current generated components directories
+		d, err := os.Open(clusterDir)
+		if err != nil {
+			log.Fatal().Err(err).Msg("")
+		}
+		defer d.Close()
+		currentCompList, err = d.Readdirnames(-1)
+		if err != nil {
+			log.Fatal().Err(err).Msg("")
+		}
 	}
 	sort.Strings(compList) // process components in sorted order
 
+	// Sort out orphaned generated components directories
+	tmpMap := make(map[string]struct{}, len(clusterComponents)) // using temp map was faster and only works when comparing two slices
+	for e, _ := range clusterComponents {
+		tmpMap[e] = struct{}{}
```

Comment:
Why is it a map?

Reply:
due to the above comment with results - https://github.com/apptio/kr8/pull/40#discussion_r758071154 

comparing slices directly somewhat processed many duplicates while looping. using map was a lot neater

Reply:
updated it to compare with slices, and turned off noisy logs